napi.h

```diff
@@ -1820,6 +1820,205 @@ namespace Napi {
+                                  const Object& resource,
+                                  ResourceString resourceName,
+                                  size_t maxQueueSize,
+                                  size_t initialThreadCount,
+                                  ContextType* context,
+                                  Finalizer finalizeCallback);
+
+    // Default set + Resource + Context + Finalizer + Data
+    template <typename ResourceString, typename ContextType,
+              typename Finalizer, typename FinalizerDataType>
+    static ThreadSafeFunction New(napi_env env,
+                                  const Function& callback,
+                                  const Object& resource,
+                                  ResourceString resourceName,
+                                  size_t maxQueueSize,
+                                  size_t initialThreadCount,
+                                  ContextType* context,
+                                  Finalizer finalizeCallback,
+                                  FinalizerDataType* data);
+
+    ThreadSafeFunction();
+    ~ThreadSafeFunction();
+
+    ThreadSafeFunction(ThreadSafeFunction&& other);
+    ThreadSafeFunction& operator=(ThreadSafeFunction&& other);
+
+    Status BlockingCall() const;
+
+    template <typename Callback>
+    Status BlockingCall(Callback callback) const;
+
+    template <typename DataType, typename Callback>
+    Status BlockingCall(DataType* data, Callback callback) const;
+
+    Status NonBlockingCall() const;
+
+    template <typename Callback>
+    Status NonBlockingCall(Callback callback) const;
+
+    template <typename DataType, typename Callback>
+    Status NonBlockingCall(DataType* data, Callback callback) const;
+
+    bool Acquire() const;
+    bool Release();
+    bool Abort();
+
+    bool IsAborted() const;
+
+    struct ConvertibleContext
+    {
+      template <class T>
+      operator T*() { return static_cast<T*>(context); }
+      void* context;
+    };
+    ConvertibleContext GetContext() const;
+
+  private:
+    using CallbackWrapper = std::function<void(Napi::Env, Napi::Function)>;
+
+    template <typename ResourceString, typename ContextType,
+              typename Finalizer, typename FinalizerDataType>
+    static ThreadSafeFunction New(napi_env env,
+                                  const Function& callback,
+                                  const Object& resource,
+                                  ResourceString resourceName,
+                                  size_t maxQueueSize,
+                                  size_t initialThreadCount,
+                                  ContextType* context,
+                                  Finalizer finalizeCallback,
+                                  FinalizerDataType* data,
+                                  napi_finalize wrapper);
+
+    ThreadSafeFunction(napi_env env, napi_threadsafe_function tsFunctionValue);
```

Comment:
This should maybe not be private, so we can allow users to mix C-API and C++-API calls. We have similar constructors for things like `Napi::Object`, which are also public.